### PR TITLE
[CAS] Introduce ActionCache into llvm::cas

### PR DESCRIPTION
**Pull request by cachemeifyoucan**
Introduce ActionCache which can be used together with a CAS to cache
computation results with a key.

### CODE REVIEW

**Comment by cachemeifyoucan** [comment]
Finish the ActionCache implementation. This is an simplified version (intentionally) from Duncan's scratch: https://github.com/apple/llvm-project/pull/4715

Currently only support CASID -> CASID mapping and the key value storage is from HashValue -> HashValue. Please check interface and all the `TODO` in added the code for some of the concerns of initial version.

I will start to remove the ActionCache from CASDB and rename CASDB next.

**Comment by cachemeifyoucan** [comment]
@swift-ci please test

**Comment by cachemeifyoucan** [comment]
@swift-ci please test

File: llvm/include/llvm/CAS/ActionCache.h [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+#include "llvm/Support/Error.h"
+
+namespace llvm::cas {
+
+class CASDB;
+class CASID;
+class ObjectProxy;
+
+/// A key for caching an operation.
+/// It is implemented as a bag of bytes and provided convenient constructor
```
Comment (benlangmuir) [inline llvm/include/llvm/CAS/ActionCache.h]: > provided convenient

Nit: should be "provide**s a** convenient ..."

File: llvm/include/llvm/CAS/ActionCache.h [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+/// for string/CAS types.
+// TODO: Support CacheKey other than a CASID but rather any array of bytes.
```
Comment (benlangmuir) [inline llvm/include/llvm/CAS/ActionCache.h]: Nit: having a regular comment between the doc comment and the decl breaks the connection between them in some tools.

File: llvm/include/llvm/CAS/ActionCache.h [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+  CacheKey(const ObjectProxy &Proxy);
+  CacheKey(const CASDB &CAS, const ObjectRef &Ref);
+
+private:
+  std::string Key;
+};
+
+/// A cache from a key describing an action to the result of doing it.
+///
+/// Actions are expected to be pure (write-once, collision is an error).
```
Comment (benlangmuir) [inline llvm/include/llvm/CAS/ActionCache.h]: It's only an error if the value changes, right? Maybe call that out since "write-once" makes it sound like `put`ing the same key-value pair twice could be an error.

File: llvm/include/llvm/CAS/ActionCache.h [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    return putImpl(arrayRefFromStringRef(ActionKey.getKey()), Result);
+  }
+
+  /// Cache \p Result using \p ObjectRef as a key.
+  Error put(const ObjectRef &RefKey, const ObjectRef &Result) {
+    CacheKey ActionKey(CAS, RefKey);
+    return putImpl(arrayRefFromStringRef(ActionKey.getKey()), Result);
+  }
+
+  /// Poison the cache for \p ActionKey.
```
Comment (benlangmuir) [inline llvm/include/llvm/CAS/ActionCache.h]: This doc comment could use a few words about what "poison" means for future calls to `get` and `put`.

File: llvm/include/llvm/CAS/ActionCache.h [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+/// A cache from a key describing an action to the result of doing it.
+///
+/// Actions are expected to be pure (write-once, collision is an error).
+class ActionCache {
+  virtual void anchor();
+
+public:
+  /// Get a previously computed result for \p ActionKey.
+  Optional<ObjectRef> get(const CacheKey &ActionKey) const {
```
Comment (benlangmuir) [inline llvm/include/llvm/CAS/ActionCache.h]: Your comment above mentions that the intention is to support arbitrary byte strings as keys - what's the plan for *values*? Will they also be arbitrary bytes, or are they required to be CAS objects?

File: llvm/include/llvm/CAS/CASDB.h [hunk 6]
```diff
@@ -141,10 +141,11 @@ class CASDB : public CASIDContext {
   virtual CASID getID(ObjectHandle Handle) const = 0;
 
   /// Get a reference to the object called \p ID.
-  ///
-  /// Returns \c None if not stored in this CAS.
   virtual Optional<ObjectRef> getReference(const CASID &ID) const = 0;
```
Comment (benlangmuir) [inline llvm/include/llvm/CAS/CASDB.h:146]: Is this change about allowing `ObjectRef` for non-existent values? If so the doc comments for `CASDB` and `ObjectRef` also need updates.  Maybe that should be a separate change entirely though.

File: llvm/include/llvm/CAS/CASDB.h [hunk 7]
```diff
@@ -141,10 +141,11 @@ class CASDB : public CASIDContext {
...
 
+  /// Get a reference to the object has the hash value \p Hash.
+  virtual Optional<ObjectRef> getReference(ArrayRef<uint8_t> Hash) const = 0;
```
Comment (benlangmuir) [inline llvm/include/llvm/CAS/CASDB.h:149]: Why do we need this separately from the CASID version?

File: llvm/unittests/CAS/ActionCacheTest.cpp [hunk 8]
```diff
@@ -0,0 +1,55 @@
+//===- ActionCacheTest.cpp ------------------------------------------------===//
```
Comment (benlangmuir) [inline llvm/unittests/CAS/ActionCacheTest.cpp:1]: Some additional tests
* miss, put, hit
* put the same key-value twice

File: llvm/unittests/CAS/ActionCacheTest.cpp [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+  Optional<ObjectProxy> ID1, ID2;
+  ASSERT_THAT_ERROR(CAS->createProxy(None, "1").moveInto(ID1), Succeeded());
+  ASSERT_THAT_ERROR(CAS->createProxy(None, "2").moveInto(ID2), Succeeded());
+  ASSERT_THAT_ERROR(Cache->put(*ID1, CAS->getReference(*ID2)), Succeeded());
+  // This is a cache miss for looking up a key doesn't exist.
+  ASSERT_FALSE(Cache->get(*ID2));
+}
+
+TEST_P(CASTest, ActionCachePoison) {
```
Comment (benlangmuir) [inline llvm/unittests/CAS/ActionCacheTest.cpp]: This test is named "poison", but does not test the `poison` method, which seems odd.

File: llvm/lib/CAS/ActionCaches.cpp [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+  using DataT = CacheEntry<sizeof(HashType)>;
+  using ActionCacheResultT = std::atomic<DataT *>;
+};
+} // end namespace
+
+// TODO: Check the hash schema is the same between action cache and CAS. If we
+// can derive that from static type information, that would be even better.
+InMemoryActionCache::InMemoryActionCache(CASDB &CAS) : ActionCache(CAS) {}
+
+OnDiskActionCache::OnDiskActionCache(CASDB &CAS, StringRef Path,
```
Comment (benlangmuir) [inline llvm/lib/CAS/ActionCaches.cpp]: Style nit: maybe group `OnDiskActionCache` and `InMemoryActionCache` methods rather than interleaving them?

File: llvm/lib/CAS/ActionCaches.cpp [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+  auto Result = Cache.find(Key);
+  if (!Result)
+    return None;
+  return getCAS().getReference(Result->Data.getValue());
+}
+
+static Error createResultCachePoisonedError(StringRef Key, StringRef Output,
+                                            StringRef ExistingOutput) {
+  return createStringError(std::make_error_code(std::errc::invalid_argument),
+                           "cache poisoned for '" + Key + "' (new='" + Output +
```
Comment (benlangmuir) [inline llvm/lib/CAS/ActionCaches.cpp:112]: This terminology is a bit confusing: we have a `poison` method  that presumably mutates the cache, but we are also returning an error for "poisoning" when the cache has not been mutated (the existing value is left in place).

Also, "poison" is not a great term for this: in the context of the compiler it's more likely this represents non-determinism not a deliberate cache poisoning attack.  I would reword this to something like "cannot modify existing value for key".

File: llvm/lib/CAS/ActionCaches.cpp [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+                                   const ObjectRef &Result) {
+  DataT Expected(getCAS().getID(Result).getHash());
+  const InMemoryCacheT::value_type &Cached = *Cache.insertLazy(
+      Key, [&](auto ValueConstructor) { ValueConstructor.emplace(Expected); });
+
+  const DataT &Observed = Cached.Data;
+  if (Expected.getValue() == Observed.getValue())
+    return Error::success();
+
+  return createResultCachePoisonedError(toStringRef(Key),
```
Comment (benlangmuir) [inline llvm/lib/CAS/ActionCaches.cpp]: Is the intention that we would call `poison` here once it is implemented? Worth a TODO if so.

File: llvm/lib/CAS/ActionCaches.cpp [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+  // value;
+  return createStringError(std::errc::not_supported, "Not implemented");
+}
+
+Expected<std::unique_ptr<OnDiskActionCache>>
+OnDiskActionCache::create(CASDB &CAS, StringRef AbsPath) {
+  if (std::error_code EC = sys::fs::create_directories(AbsPath))
+    return createFileError(AbsPath, EC);
+
+  const StringRef Slash = sys::path::get_separator();
```
Comment (benlangmuir) [inline llvm/lib/CAS/ActionCaches.cpp]: How about using `sys::path::append` instead?

File: llvm/include/llvm/CAS/ActionCache.h [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+
+private:
+  CASDB &CAS;
+};
+
+/// Create an action cache in memory.
+std::unique_ptr<ActionCache> createInMemoryActionCache(CASDB &CAS);
+
+/// Create an action cache on disk.
+Expected<std::unique_ptr<ActionCache>> createOnDiskActionCache(CASDB &CAS,
```
Comment (benlangmuir) [inline llvm/include/llvm/CAS/ActionCache.h:94]: Is it supported to use the same Path for the action cache and the CAS? I would hate to have two separate command-line arguments for these paths.

File: llvm/include/llvm/CAS/ActionCache.h (hunk 4 above)
Comment (cachemeifyoucan) [reply llvm/include/llvm/CAS/ActionCache.h]: I am still thinking what this should mean. You can see it is currently "unimplemented". Ideas?

I am currently leaning towards a method to remove/reset the entry. Maybe "poison" is not the correct name.

File: llvm/include/llvm/CAS/ActionCache.h (hunk 5 above)
Comment (cachemeifyoucan) [reply llvm/include/llvm/CAS/ActionCache.h]: TBD but I am leaning towards always requiring CAS objects since it is a more efficient storage and can avoid unnecessary "poison" entries.

Comment (akyrtzi) [reply llvm/include/llvm/CAS/ActionCache.h]: >  I am leaning towards always requiring CAS objects since it is a more efficient storage

Agreed.

File: llvm/include/llvm/CAS/CASDB.h (hunk 7 above)
Comment (cachemeifyoucan) [reply llvm/include/llvm/CAS/CASDB.h:149]: We don't have a method to construct `CASID` from ArrayRef of bytes and ArrayRef is how it is stored in ActionCache.

File: llvm/include/llvm/CAS/ActionCache.h (hunk 4 above)
Comment (benlangmuir) [reply llvm/include/llvm/CAS/ActionCache.h]: If it will remove the entry, I think `remove` is fine.  "Poison" to me implies that it will never be valid to use this key again.

File: llvm/include/llvm/CAS/ActionCache.h (hunk 14 above)
Comment (cachemeifyoucan) [reply llvm/include/llvm/CAS/ActionCache.h:94]: Yes, even though it is default to different directory so you can remove one not the other. If you put into the same directory, you need to know what to remove so you don't leave the system in an invalid state.

Adding a patch to see how integration works (it is more work to pass two variables down into all the tools).

File: llvm/include/llvm/CAS/ActionCache.h (hunk 4 above)
Comment (cachemeifyoucan) [reply llvm/include/llvm/CAS/ActionCache.h]: I will delete this method now before we figure out what to do.

**Comment by cachemeifyoucan** [comment]
@swift-ci please test

File: llvm/lib/CAS/ActionCaches.cpp (hunk 12 above)
Comment (cachemeifyoucan) [reply llvm/lib/CAS/ActionCaches.cpp]: I kept this name for now since the `poison` method is removed for now.

File: llvm/include/llvm/CAS/ActionCache.h (hunk 1 above)
Comment (benlangmuir) [reply llvm/include/llvm/CAS/ActionCache.h]: I think you missed that "d" should be "s" in "provides".

File: llvm/include/llvm/CAS/ActionCache.h (hunk 5 above)
Comment (benlangmuir) [reply llvm/include/llvm/CAS/ActionCache.h]: SGTM.

File: llvm/lib/CAS/ActionCaches.cpp [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+  const DataT *Observed = reinterpret_cast<const DataT *>(ActionP->Data.data());
+
+  if (Expected.getValue() == Observed->getValue())
+    return Error::success();
+
+  return createResultCachePoisonedError(hashToString(Key), getCAS(), Result,
+                                        Observed->getValue());
+}
+
+static constexpr StringLiteral DefaultName = "llvm.actioncache.builtin.default";
```
Comment (benlangmuir) [inline llvm/lib/CAS/ActionCaches.cpp]: How about putting the cas and cache into a common parent directory?

File: clang/include/clang/CAS/CASOptions.h [hunk 16]
```diff
@@ -73,7 +76,7 @@ class CASConfiguration {
 /// defined in CASConfiguration to enable caching a CAS instance.
 ///
 /// CASOptions includes \a getOrCreateCAS() and \a
-/// getOrCreateCASAndHideConfig() for creating a CAS and caching it.
+/// getOrCreateActionCache() for creating CAS and ActionAction.
```
Comment (benlangmuir) [inline clang/include/clang/CAS/CASOptions.h]: Typo: "ActionAction"

File: clang/tools/driver/CacheLauncherMode.cpp [hunk 17]
```diff
@@ -144,6 +144,8 @@ clang::handleClangCacheInvocation(SmallVectorImpl<const char *> &Args,
     }
     if (const char *CASPath = ::getenv("LLVM_CACHE_CAS_PATH")) {
       Args.append({"-Xclang", "-fcas-path", "-Xclang", CASPath});
+      // Put ActionCache in the same directory as CAS.
```
Comment (benlangmuir) [inline clang/tools/driver/CacheLauncherMode.cpp]: How about creating sub-directories to keep the files separate?

File: llvm/include/llvm/CAS/ActionCache.h [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+
+/// A cache from a key describing an action to the result of doing it.
+///
+/// Actions are expected to be pure (collision is an error).
+class ActionCache {
+  virtual void anchor();
+
+public:
+  /// Get a previously computed result for \p ActionKey.
+  Optional<ObjectRef> get(const CacheKey &ActionKey) const {
```
Comment (akyrtzi) [inline llvm/include/llvm/CAS/ActionCache.h]: (putting some offline feedback in the PR)

The `get()` API should return `Expected<Optional<ObjectRef>>`. The client should be able to distinguish between "the key is missing" and "there was some error accessing the ActionCache" scenarios.

**Comment by cachemeifyoucan** [comment]
@swift-ci please test

**Comment by cachemeifyoucan** [comment]
@swift-ci please test

**Comment by cachemeifyoucan** [comment]
@swift-ci please test

**Comment by cachemeifyoucan** [comment]
@swift-ci please test

File: llvm/include/llvm/CAS/ActionCache.h [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+
+protected:
+  virtual Expected<Optional<ObjectRef>>
+  getImpl(ArrayRef<uint8_t> ResolvedKey) const = 0;
+  virtual Error putImpl(ArrayRef<uint8_t> ResolvedKey,
+                        const ObjectRef &Result) = 0;
+
+  ActionCache(CASDB &CAS) : CAS(CAS) {}
+
+  CASDB &getCAS() const { return CAS; }
```
Comment (benlangmuir) [inline llvm/include/llvm/CAS/ActionCache.h:80]: If we made this public, we could potentially avoid needing to pass both a CAS and an ActionCache everywhere.  It seems like to use the cache you are already tied to using a specific CAS since otherwise the ObjectRef won't make sense.  WDYT?

File: llvm/include/llvm/CAS/CachingOnDiskFileSystem.h [hunk 20]
```diff
@@ -110,19 +113,23 @@ class CachingOnDiskFileSystem : public ThreadSafeFileSystem {
   }
 
 protected:
-  CachingOnDiskFileSystem(std::shared_ptr<CASDB> DB);
-  CachingOnDiskFileSystem(CASDB &DB);
+  CachingOnDiskFileSystem(std::shared_ptr<CASDB> DB,
+                          std::shared_ptr<ActionCache> Cache);
+  CachingOnDiskFileSystem(CASDB &DB, ActionCache &Cache);
```
Comment (benlangmuir) [inline llvm/include/llvm/CAS/CachingOnDiskFileSystem.h]: Why are we passing the cache into the filesystem? It doesn't seem to be used in the .cpp file except to pass it through the constructor/etc.

File: llvm/include/llvm/CAS/ActionCache.h (hunk 19 above)
Comment (cachemeifyoucan) [reply llvm/include/llvm/CAS/ActionCache.h:80]: That doesn't really work unless we make ActionCache owns CAS, otherwise this is going to open to all kinds of life time issues.

File: llvm/include/llvm/CAS/CachingOnDiskFileSystem.h (hunk 20 above)
Comment (cachemeifyoucan) [reply llvm/include/llvm/CAS/CachingOnDiskFileSystem.h]: Well, the dependency scanner, grabs the CAS from FileSystem. Maybe we need a different way to create an ActionCache into dep scanner.

**Comment by cachemeifyoucan** [comment]
@swift-ci please test

**Comment by cachemeifyoucan** [comment]
@swift-ci please test

**Comment by cachemeifyoucan** [comment]
@swift-ci  please test

**Comment by cachemeifyoucan** [comment]
@swift-ci please test

File: llvm/include/llvm/CAS/CachingOnDiskFileSystem.h [hunk 21]
```diff
@@ -14,10 +14,12 @@
 #include "llvm/Support/Error.h"
 #include "llvm/Support/VirtualFileSystem.h"
 #include <map>
+#include <memory>
 
 namespace llvm {
 namespace cas {
 
+class ActionCache;
```
Comment (benlangmuir) [inline llvm/include/llvm/CAS/CachingOnDiskFileSystem.h]: Looks like there are a couple left-over changes in this file (and the .cpp as well).

**Review by benlangmuir — APPROVED** [review]

**Comment by cachemeifyoucan** [comment]
@swift-ci please test
